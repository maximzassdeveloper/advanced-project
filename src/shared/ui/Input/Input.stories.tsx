import { Meta } from '@storybook/react'
import { Input } from './Input'

const meta: Meta<typeof Input> = {
  title: 'shared/Input',
  component: Input,
}
export default meta

export const Default = () => (
  <div className='flex column'>
    <Input placeholder='Wow placeholder' />
    <Input value='Woow value' />
  </div>
)

export const WithLabel = () => (
  <div>
    <Input value='Woow value' label='Label for input' />
  </div>
)

export const WithError = () => (
  <div className='flex column'>
    <Input value='Woow value' error />
    <Input value='Woow value' error='error message' />
  </div>
)

export const Readonly = () => (
  <div>
    <Input value='Woow value' readOnly />
  </div>
)
