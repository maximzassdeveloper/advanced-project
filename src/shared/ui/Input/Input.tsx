import { forwardRef, InputHTMLAttributes, memo } from 'react'
import { Typography } from '@/shared/ui'
import { classNames } from '@/shared/lib/classNames'
import s from './input.module.scss'

interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  className?: string
  error?: boolean | string
  label?: string
}

export const Input = memo(
  forwardRef<HTMLInputElement, InputProps>((props, ref) => {
    const { className, id, autoComplete = 'off', error, label, readOnly, ...rest } = props

    const classes = classNames(s.input, className, {
      [s.errored]: !!error,
      [s.readonly]: readOnly,
    })

    return (
      <div className={classes} data-testid='input'>
        {!!label && (
          <label htmlFor={id} className={s.label}>
            {label}
          </label>
        )}

        <input ref={ref} id={id} autoComplete={autoComplete} readOnly={readOnly} {...rest} />

        {typeof error === 'string' && (
          <Typography.Text
            data-testid='error'
            className={s.error}
            theme='error'
            weight='500'
            size='xs'
          >
            {error}
          </Typography.Text>
        )}
      </div>
    )
  })
)
