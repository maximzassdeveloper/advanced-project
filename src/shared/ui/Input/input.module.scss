.input {
  position: relative;
  width: 100%;
  max-width: 400px;
  transition: all 0.2s;

  &.readonly {
    opacity: 0.8;

    input:focus {
      outline: none !important;
    }
  }

  &.errored {
    input {
      border-color: var(--danger);
    }
  }

  &:not(.readonly):not(.errored) {
    // input hover
    input:not(:disabled):hover {
      border-color: var(--main);
    }

    // input focus
    input:focus {
      border-color: var(--main);

      // box-shadow: 0 0 0 2px var(--main);
    }
  }

  input {
    width: 100%;
    padding: 9px 14px;
    font: var(--font-s);
    font-weight: 700;
    color: var(--main);
    background: transparent;
    border: 1px solid var(--secondary);
    border-radius: 3px;
    transition: all 0.2s;
    outline: none !important;

    @include focus-visible-outline;

    &::placeholder {
      color: var(--light-grey);
    }

    &:disabled {
      pointer-events: none;
      opacity: 0.8;
    }
  }
}

.label {
  display: inline-block;
  font: var(--font-xs);
  font-weight: 500;
  color: var(--secondary);
  margin-bottom: 5px;
}

.error {
  margin-top: 3px;
}
