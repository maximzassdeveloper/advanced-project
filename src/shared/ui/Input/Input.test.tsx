import { render, screen } from '@testing-library/react'
import { Input } from './Input'

describe('Input', () => {
  test('should return correct snapshot', () => {
    const { container } = render(<Input label='wow label' defaultValue='wow' />)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('should show error message', () => {
    render(<Input error='Error by test' />)

    expect(screen.getByText(/Error by test/i)).toBeInTheDocument()
    expect(screen.getByTestId('input')).toMatchSnapshot()
  })

  test('should not show error element if error prop boolean', () => {
    render(<Input error />)

    expect(screen.queryByTestId('error')).toBeNull()
    expect(screen.getByTestId('input')).toHaveClass('errored')
  })

  test('should render label', () => {
    render(<Input label='input label' id='name' />)

    const label = screen.getByText(/input label/i)
    const input = screen.getByLabelText(/input label/i)

    expect(label).toBeInTheDocument()
    expect(label).toHaveAttribute('for', 'name')
    expect(input).toHaveAttribute('id', 'name')

    expect(screen.getByTestId('input')).toMatchSnapshot()
  })
})
