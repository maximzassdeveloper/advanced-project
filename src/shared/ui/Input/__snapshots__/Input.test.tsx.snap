// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render label 1`] = `
<div
  class="input"
  data-testid="input"
>
  <label
    class="label"
    for="name"
  >
    input label
  </label>
  <input
    autocomplete="off"
    id="name"
  />
</div>
`;

exports[`Input should return correct snapshot 1`] = `
<div
  class="input"
  data-testid="input"
>
  <label
    class="label"
  >
    wow label
  </label>
  <input
    autocomplete="off"
    value="wow"
  />
</div>
`;

exports[`Input should show error message 1`] = `
<div
  class="input errored"
  data-testid="input"
>
  <input
    autocomplete="off"
  />
  <p
    class="text error xs weight-500 error"
    data-testid="error"
  >
    Error by test
  </p>
</div>
`;
